### PR TITLE
UOE-9785: Custom Dimensions Support for In-App

### PR DESCRIPTION
**Pull request by pm-saurabh-narkhede**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: modules/pubmatic/openwrap/auctionresponsehook.go [hunk 1]
```diff
@@ -110,6 +111,16 @@ func (m OpenWrap) handleAuctionResponseHook(
 
 			if bidExt.Prebid != nil {
 				bidExt.CreativeType = string(bidExt.Prebid.Type)
+				if bidExt.Prebid.Targeting == nil {
+					bidExt.Prebid.Targeting = make(map[string]string)
+				}
+				if cdsMap, ok := customdimensions.IsCustomDimensionsPresent(rctx.NewReqExt); ok {
+					for key, value := range cdsMap {
+						if value.SendToGAM == nil || value.SendToGAM != nil && *value.SendToGAM {
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/auctionresponsehook.go:119]: Can be rewritten like value.SendToGAM == nil || (value.SendToGAM != nil && *value.SendToGAM) for more redablity.  please add the comment as well.

File: modules/pubmatic/openwrap/customdimensions/customdimension_test.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+				"age": {
+					Value: "23",
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := GetCustomDimensionsFromRequestExt(tt.args.bidderParams); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("GetCustomDimensionsFromRequestExt() = %v, want %v", got, tt.want)
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/customdimensions/customdimension_test.go:58]: can we use assert.equal(), more easy for debugging.

File: modules/pubmatic/openwrap/customdimensions/customdimension_test.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+				cdsMap: map[string]CustomDimension{},
+			},
+			want:  ``,
+			want2: ``,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := ParseCustomDimensionsToString(tt.args.cdsMap); got != tt.want && got != tt.want2 {
+				t.Errorf("ParseCustomDimensionsToString() = %v, want %v", got, tt.want)
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/customdimensions/customdimension_test.go:97]: can we use assert.equal(), more easy for debugging.

File: modules/pubmatic/openwrap/customdimensions/customdimension_test.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+				ext: nil,
+			},
+			want:  map[string]CustomDimension{},
+			want1: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, got1 := IsCustomDimensionsPresent(tt.args.ext)
+			if !reflect.DeepEqual(got, tt.want) {
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/customdimensions/customdimension_test.go:167]: can we use assert.equal() istead of reflect.DeepEqual(), more easy for debugging.

File: modules/pubmatic/openwrap/auctionresponsehook.go (hunk 1 above)
Comment (pm-saurabh-narkhede) [reply modules/pubmatic/openwrap/auctionresponsehook.go]: Done

File: modules/pubmatic/openwrap/customdimensions/customdimension_test.go (hunk 4 above)
Comment (pm-saurabh-narkhede) [reply modules/pubmatic/openwrap/customdimensions/customdimension_test.go]: Done

**Review by AvinashKapre — APPROVED** [review]

File: analytics/pubmatic/logger.go [hunk 5]
```diff
@@ -134,8 +134,9 @@ func GetLogAuctionObjectAsURL(ao analytics.AuctionObject, rCtx *models.RequestCt
 		for i := range wlog.Slots {
 			wlog.Slots[i].FloorSkippedFlag = tracker.Tracker.FloorSkippedFlag
 		}
```
Comment (ashishshinde-pubm) [inline analytics/pubmatic/logger.go:136]: plz check whether we should log 'cds' if trackers are absent. (winning bid is not present)

File: modules/pubmatic/openwrap/customdimensions/customdimension_test.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+func TestConvertCustomDimensionsToString(t *testing.T) {
+	type args struct {
+		cdsMap map[string]models.CustomDimension
+	}
+	tests := []struct {
+		name  string
+		args  args
+		want  string
+		want2 string
```
Comment (ashishshinde-pubm) [inline modules/pubmatic/openwrap/customdimensions/customdimension_test.go:73]: plz remove this

File: analytics/pubmatic/logger.go (hunk 5 above)
Comment (pm-saurabh-narkhede) [reply analytics/pubmatic/logger.go:138]: Added cds incase trackers are abset

File: modules/pubmatic/openwrap/customdimensions/customdimension_test.go (hunk 6 above)
Comment (pm-saurabh-narkhede) [reply modules/pubmatic/openwrap/customdimensions/customdimension_test.go]: done

**Review by pm-nilesh-chate — APPROVED** [review]

**Review by AvinashKapre — APPROVED** [review]

**Pull request merged**